File: src/main/java/io/bloco/faker/components/Date.java
package io.bloco.faker.components;

import org.joda.time.DateTime;
import org.joda.time.Duration;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;

import io.bloco.faker.FakerComponent;
import io.bloco.faker.FakerData;

public class Date extends FakerComponent {

    private static final int DEFAULT_NUM_OF_DAYS = 365;
    private static final int DEFAULT_MIN_AGE = 18;
    private static final int DEFAULT_MAX_AGE = 65;

    private final DateFormat dateFormat;

    public Date(FakerData fakerData) {
        super(fakerData);
        dateFormat = new SimpleDateFormat("yyyy-MM-dd");
    }

    public java.util.Date between(String from, String to) {
        java.util.Date fromDate, toDate;

        try {
            fromDate = dateFormat.parse(from);
            toDate = dateFormat.parse(to);
        } catch (ParseException exception) {
            throw new RuntimeException(exception);
        }

        return between(fromDate, toDate);
    }

    public java.util.Date between(java.util.Date from, java.util.Date to) {
        return between(new DateTime(from), new DateTime(to));
    }

    public java.util.Date forward() {
        return forward(DEFAULT_NUM_OF_DAYS);
    }

    public java.util.Date forward(int numberOfDays) {
        return between(today(), today().plusDays(numberOfDays));
    }

    public java.util.Date backward() {
        return backward(DEFAULT_NUM_OF_DAYS);
    }

    public java.util.Date backward(int numberOfDays) {
        return between(today().minusDays(numberOfDays), today());
    }

    public java.util.Date birthday() {
        return birthday(DEFAULT_MIN_AGE, DEFAULT_MAX_AGE);
    }

    public java.util.Date birthday(int minAge, int maxAge) {
        return between(today().plusYears(minAge), today().plusYears(maxAge));
    }

    // Helpers

    private java.util.Date between(DateTime from, DateTime to) {
        return getRandomDate(from, to).withTimeAtStartOfDay().toDate();
    }

    private DateTime today() {
        return new DateTime().withTimeAtStartOfDay();
    }

    private DateTime getRandomDate(DateTime from, DateTime to) {
        Duration duration = new Duration(from, to);
        long diffInDays = duration.getStandardDays();
        return from.plusDays((int) randomHelper.number(diffInDays) + 1);
    }
}

File: lib/src/main/kotlin/io/bloco/faker/components/Date.kt
package io.bloco.faker.components

import io.bloco.faker.FakerComponent
import io.bloco.faker.FakerData
import io.bloco.faker.helpers.toDate
import io.bloco.faker.helpers.toLocalDate
import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.util.Date
import kotlin.random.Random

class Date(fakerData: FakerData) : FakerComponent(fakerData) {
    private val formatter = SimpleDateFormat("yyyy-MM-dd")

    fun between(from: String, to: String): Date {
        return try {
            val fromDate = formatter.parse(from)
            val toDate = formatter.parse(to)
            between(fromDate, toDate)
        } catch (exception: ParseException) {
            throw RuntimeException(exception)
        }
    }

    fun between(from: Date, to: Date): Date {
        return between(from.toLocalDate(), to.toLocalDate())
    }

    @JvmOverloads
    fun forward(numberOfDays: Int = DEFAULT_NUM_OF_DAYS): Date {
        return between(today(), today().plusDays(numberOfDays.toLong()))
    }

    @JvmOverloads
    fun backward(numberOfDays: Int = DEFAULT_NUM_OF_DAYS): Date {
        return between(today().minusDays(numberOfDays.toLong()), today())
    }

    @JvmOverloads
    fun birthday(minAge: Int = DEFAULT_MIN_AGE, maxAge: Int = DEFAULT_MAX_AGE): Date {
        return between(today().plusYears(minAge.toLong()), today().plusYears(maxAge.toLong()))
    }

    // Helpers
    private fun between(from: LocalDate, to: LocalDate): Date {
        return getRandomDate(from, to)
    }

    private fun today(): LocalDate {
        return LocalDate.now()
    }

    private fun getRandomDate(from: LocalDate, to: LocalDate): Date {
        val diffInDays = ChronoUnit.DAYS.between(from, to)
        return from.plusDays(Random.nextLong(diffInDays)).toDate()
    }

    companion object {
        private const val DEFAULT_NUM_OF_DAYS = 365
        private const val DEFAULT_MIN_AGE = 18
        private const val DEFAULT_MAX_AGE = 65
    }
}